crates/uv-dev/src/fetch_python.rs

```diff
@@ -0,0 +1,165 @@
+use fs_err as fs;
+use std::collections::HashMap;
+use std::{path::PathBuf, str::FromStr};
+
+use anyhow::Result;
+use clap::Parser;
+
+use futures::StreamExt;
+
+use itertools::Itertools;
+use tokio::io::AsyncReadExt;
+use tokio::{fs::File, time::Instant};
+
+#[cfg(windows)]
+use fs_err::tokio::hard_link;
+#[cfg(unix)]
+use fs_err::tokio::symlink;
+
+use tracing::{info, info_span, Instrument};
+
+use uv_fs::Simplified;
+use uv_toolchain::{DownloadResult, Error, PythonDownload, PythonDownloadRequest};
+
+#[derive(Parser, Debug)]
+pub(crate) struct FetchPythonArgs {
+    versions: Vec<String>,
+}
+
+pub(crate) async fn fetch_python(args: FetchPythonArgs) -> Result<()> {
+    let start = Instant::now();
+
+    let bootstrap_dir = std::env::var_os("UV_BOOTSTRAP_DIR")
+        .map(PathBuf::from)
+        .unwrap_or(std::env::current_dir()?.join("bin"));
+
+    fs_err::create_dir_all(&bootstrap_dir)?;
+
+    let versions = if args.versions.is_empty() {
+        info!("Reading versions from file...");
+        read_versions_file().await?
+    } else {
+        args.versions
+    };
+
+    let requests = versions
+        .iter()
+        .map(|version| match PythonDownloadRequest::from_str(version) {
+            Ok(request) => request.fill(),
+            err @ Err(_) => err,
+        })
+        .collect::<Result<Vec<_>, Error>>()?;
+
+    let downloads = requests
+        .iter()
+        .map(|request| match PythonDownload::from_request(request) {
+            Some(download) => download,
+            None => panic!("No download found for request {request:?}"),
+        })
+        .collect::<Vec<_>>();
+
+    let client = uv_client::BaseClientBuilder::new().build();
+
+    info!("Fetching requested versions...");
+    let mut tasks = futures::stream::iter(downloads.iter())
+        .map(|download| {
+            async {
+                let result = download.fetch(&client, &bootstrap_dir).await;
+                (download.python_version(), result)
+            }
+            .instrument(info_span!("download", key = %download))
+        })
+        .buffered(4);
+
+    let mut results = Vec::new();
+    let mut downloaded = 0;
+    while let Some(task) = tasks.next().await {
+        let (version, result) = task;
+        let path = match result? {
+            DownloadResult::AlreadyAvailable(path) => {
+                info!("Found existing download for v{}", version);
+                path
+            }
+            DownloadResult::Fetched(path) => {
+                info!("Downloaded v{} to {}", version, &path.user_display());
+                downloaded += 1;
```

Comment:
Why this change?

Reply:
The `&` is not required, `format!` always borrows the value. Normally clippy would flag this (`needless_borrow`), but apparently it can't look through this macro.

Reply:
Thanks! Got confused by the `downloaded` removal.